File: apps/dsa-web/src/api/tiered.ts
import apiClient from './index';

// Types mirror the backend JSON verbatim (snake_case). Dimension payloads
// hold metric names like revenue_yoy_pct — do not camelize them.
export type TieredCitation = {
  source_name: string;
  url: string | null;
  title: string | null;
  snippet: string | null;
};

export type TieredDimension = {
  dimension: string;
  kind: 'numeric' | 'textual';
  coverage: 'full' | 'partial' | 'unavailable';
  is_actionable: boolean;
  payload: Record<string, unknown> | null;
  narrative: string | null;
  warnings: string[];
  citations: TieredCitation[];
};

export type TieredLevels = {
  entry: number | null;
  secondary_entry: number | null;
  stop_loss: number | null;
  take_profit: number | null;
};

// v2 slice 3 audit trail: per-level formula base + validated AI adjustment.
// The plan-review redesign (2026-07-22) adds a "shares" entry in the same
// shape (base = count from the computed levels; adjusted_inputs = the
// final levels a mechanical recompute used) and `links` — inline
// citations for the adjustment reason, in the debate-link shape.
// One adjustment reason: the flagged check it fixes (a fixed keyword id
// the UI translates) plus one cited sentence.
export type TieredLevelReason = {
  check: string;
  text: string;
  links?: TieredDebateLink[];
};

export type TieredLevelDetail = {
  base: number | null;
  formula: string | null;
  inputs: Record<string, number> | null;
  adjusted: number | null;
  /** Old stored runs: one paragraph. New runs carry `reasons` instead. */
  reason?: string | null;
  reasons?: TieredLevelReason[] | null;
  evidence: string[];
  links?: TieredDebateLink[];
  adjusted_inputs?: Record<string, number> | null;
  /** Shares only: the mechanical recompute from the adjusted levels,
   *  before any AI trim — the receipt's result line. */
  mechanical?: number | null;
  rejection: string | null;
  final: number | null;
};

// One round of the check-adjust cycle that ended with a risk check still
// firing. Present (non-empty) only when the review never converged and
// the computed plan was kept — the UI words it in the blue-keep modal.
export type TieredReviewFailure = {
  round: number;
  checks: string[];
};

export type TieredLevelsDetail = {
  levels: Record<string, TieredLevelDetail>;
  warnings: string[];
  review_failures?: TieredReviewFailure[] | null;
};

// One structured trade-plan warning: numbers only, the frontend words it.
export type TieredPlanWarning = {
  id: string;
  values: Record<string, unknown>;
};

// Per-column warning lists for the plan table's Warnings row.
export type TieredPlanWarnings = Record<string, TieredPlanWarning[]>;

export type TieredAnchoredReason = {
  claim: string;
  evidence: string[];
};

// v4 judge grade for one validity axis: the 0-5 score plus, below 5, the
// exact offending sentence and why it is wrong (both null at 5/5 → N/A).
export type TieredAxisGrade = {
  score: number;
  quote?: string | null;
  why?: string | null;
};

// One debater's numbers — its own 0-10 position score plus the judge's
// three 0-5 validity grades; weight = (sum of grades) / 15. v3 runs store
// the position score as `bullishness` and the axes as bare numbers; v4
// runs store `position_score` and axis objects with quote/why comments.
export type TieredDebaterScore = {
  position_score?: number;
  bullishness?: number;
  citation_validity: number | TieredAxisGrade;
  knowledge_validity: number | TieredAxisGrade;
  logical_validity: number | TieredAxisGrade;
  weight: number;
  notes?: string | null;
};

// --- v5/v6/v7 debate tree (defender/attacker/judge) ---

// Inline citation. v6: text = the words underlined in the claim, value =
// the claimed number (mismatch marks a code-detected contradiction).
// v7: payload links carry value (the report's display string; the claim
// contains it verbatim) with text null; sentiment links carry text (the
// words resting on that news source) with value null.
export type TieredDebateLink = {
  text?: string | null;
  ref: string;
  value: number | string | null;
  mismatch?: boolean;
};

// One citation-or-logic check; 'invalid' carries the reason + citations.
export type TieredDebateCheck = {
  verdict: 'valid' | 'invalid';
  reason: string | null;
  citations: string[];
};

// The defender's response to one challenge: its own checks ON the
// attack/addition. v6 stores the citation/logic pair (both valid →
// accepted); v7 stores the single `check` (valid → accepted). All three
// fields optional so both generations type-check.
export type TieredDebateResponse = {
  accepted: boolean;
  citation_check?: TieredDebateCheck;
  logic_check?: TieredDebateCheck;
  check?: TieredDebateCheck;
};

// The judge's word on one axis of a defender item: its own check when the
// axis was unattacked, a ruling on the attack when it was.
export type TieredDebateJudgeAxis = {
  kind: 'reason_check' | 'attack_ruling';
  verdict: 'valid' | 'invalid' | 'attack_right' | 'attack_wrong';
  reason: string | null;
  citations: string[];
};

export type TieredDebateJudgeAddition = {
  kind: 'addition_ruling';
  verdict: 'real' | 'bogus';
  reason: string | null;
  citations: string[];
};

// One v8 vote on a bullet: the check round's second vote or the
// deciding round's tiebreaker. Reasons carry the same code-verified
// links the bullets use. v10 votes carry the voter's own importance
// rating of the bullet (1-3; 1-5 since v11); v11 adds one plain
// sentence saying why (weight_reason).
export type TieredDebateVote = {
  role: 'checker' | 'decider';
  verdict: 'valid' | 'invalid';
  reason: string | null;
  links: TieredDebateLink[];
  weight?: number;
  weight_reason?: string | null;
};

// v11: one lister's own rating of a bullet they authored — which lister
// (1 or 2), the 1-5 importance score, and why.
export type TieredDebateAuthorVote = {
  lister: number;
  weight: number;
  weight_reason?: string | null;
};

// One evidence item of the tree with everything that happened to it.
// v5 runs store citations + the count/outcome ledger; v6 runs store
// links + value_check + the per-axis checks; v7 runs store single-axis
// fields (attacker_check, one response, one judge line) plus struck +
// problems for bullets whose citations code could not fix; v8 runs
// store authors (how many analysts listed it independently) + votes —
// all optional so every generation renders.
export type TieredDebateItem = {
  id: string;
  dimension: string;
  direction: 'bullish' | 'bearish';
  claim: string;
  citations?: string[];
  links?: TieredDebateLink[];
  value_check?: { verdict: 'valid' | 'invalid'; problems: string[] } | null;
  struck?: boolean;
  problems?: string[];
  authors?: number;
  votes?: TieredDebateVote[];
  // v10 weighted votes: the authors' own ratings (one or two), and the
  // final median of every voter's rating (null on struck bullets).
  // v11 adds author_votes: the same ratings with the lister number and
  // the reason attached.
  author_weights?: number[];
  author_votes?: TieredDebateAuthorVote[];
  weight?: number | null;
  added_by_attacker?: boolean;
  attacker_checks?: { citation: TieredDebateCheck; logic: TieredDebateCheck } | null;
  attacker_check?: TieredDebateCheck | null;
  responses?: {
    citation: TieredDebateResponse | null;
    logic: TieredDebateResponse | null;
  };
  response: TieredDebateResponse | null;
  judge:
    | { citation: TieredDebateJudgeAxis; logic: TieredDebateJudgeAxis }
    | TieredDebateJudgeAddition
    | TieredDebateJudgeAxis
    | null;
  count?: { numerator: number; denominator: number } | null;
  outcome?: 'valid' | 'invalid' | 'neutral' | null;
  final_status?: 'counted' | 'excluded' | null;
  exclusion_reason?: string | null;
};

// Pool snapshot: per-dimension bullish/bearish counts plus the
// pool-wide totals and score. v6-v8 store a per-dimension score and an
// averaged overall score; v9 stores flat counting (10 × bullish/total)
// and no per-dimension score; v10 adds importance-weight sums and a
// weighted score (10 × bullish weight / total weight).
export type TieredDebatePool = {
  dimensions: Record<
    string,
    {
      bullish: number;
      bearish: number;
      total: number;
      score?: number | null;
      bullish_weight?: number;
      bearish_weight?: number;
      total_weight?: number;
    }
  >;
  bullish: number;
  bearish: number;
  total: number;
  bullish_weight?: number;
  bearish_weight?: number;
  total_weight?: number;
  score: number | null;
};

// Bull/bear debate audit trail (tier-2 section). Four generations coexist
// in stored runs: the v2 judged shape (confidence, reasons, would_change_
// One bullet of the structured deep-analysis report; children are
// sub-bullets one level deep.
export type TieredSummaryBullet = { text: string; children?: string[] | null };

// The fixed report outline (owner decision 2026-07-24): the group set
// and order never change run to run — the AI only fills the bullets.
export type TieredSummaryStructure = {
  summary?: TieredSummaryBullet[] | null;
  technicals?: TieredSummaryBullet[] | null;
  fundamentals?: TieredSummaryBullet[] | null;
  positioning?: TieredSummaryBullet[] | null;
  macro_econ?: TieredSummaryBullet[] | null;
};

// mind), the v3 scored shape (final_score, scoring, corrected bull/bear
// summaries), the v4 threaded shape (turn kinds, axis-grade comments),
// and the v5 tree (format: 5, items, weight ledger) — every
// generation-specific field is optional.
export type TieredDebateDetail = {
  // 5/6/7 on tree-format runs; absent on everything stored before.
  format?: number;
  items?: TieredDebateItem[];
  turns: {
    role: string;
    // v2/v3 runs number their rounds; v4 turns carry a kind instead.
    round?: number;
    kind?: string;
    argument: string;
    bullishness?: number | null;
    position_score?: number | null;
    citations?: string[];
  }[];
  verdict: {
    direction: string;
    summary: string;
    // v11+ structured report: the fixed five-group outline the summary
    // stage fills (summary + one group per dimension); `summary` above
    // is its flat-text rendering for older readers.
    summary_structure?: TieredSummaryStructure | null;
    // v2 judged shape
    confidence?: number | null;
    reasons_for?: TieredAnchoredReason[];
    reasons_against?: TieredAnchoredReason[];
    would_change_mind?: string | null;
    // v3 scored shape
    final_score?: number | null;
    final_score_rounded?: number | null;
    bull_summary?: string | null;
    bear_summary?: string | null;
    scoring?: { bull: TieredDebaterScore; bear: TieredDebaterScore } | null;
    // v5/v6 tree shape (v5 scores are whole numbers + weight; v6 scores
    // are 2-decimal pool counts + pools)
    initial_score?: number | null;
    adjusted_score?: number | null;
    adjusted_kept?: boolean | null;
    weight?: { numerator: number; denominator: number; value: number } | null;
    // v8 pools drop the adjusted snapshot (no concede/adopt step).
    pools?: {
      initial: TieredDebatePool;
      adjusted?: TieredDebatePool;
      final: TieredDebatePool;
    } | null;
  } | null;
  warnings: string[];
};

// One tier-3 risk bullet (risk_detail format 2): the tier-2 vote-item
// shape minus the direction tag — every bullet is a risk, and code maps
// the confirmed count to the size multiplier.
export type TieredRiskItem = {
  id: string;
  dimension: string;
  claim: string;
  links?: TieredDebateLink[];
  struck?: boolean;
  problems?: string[];
  authors?: number;
  votes?: TieredDebateVote[];
  final_status?: 'counted' | 'excluded' | null;
  exclusion_reason?: string | null;
};

// Per-group and total risk counts for one pool snapshot.
export type TieredRiskCounts = {
  groups: Record<string, number>;
  total: number;
};

// Risk stress audit trail (tier-3 section). Two generations coexist in
// stored runs: the persona/judge shape (takes + stance/stop advice) and
// the format-2 risk vote (items + count-derived multiplier) — every
// generation-specific field is optional.
export type TieredRiskDetail = {
  // 2 on risk-vote runs; absent on the stored persona/judge runs.
  format?: number;
  takes: { persona: string; assessment: string }[];
  items?: TieredRiskItem[];
  verdict: {
    stance: string;
    size_multiplier: number;
    // Absent on runs stored before the risk judge reported its own 0-1
    // sureness; the UI hides the score for those.
    confidence?: number | null;
    stop_advice: string;
    tightened_stop: number | null;
    summary: string;
    key_risks: TieredAnchoredReason[];
    // format 2: the code-owned count → multiplier arithmetic.
    confirmed_risks?: number;
    total_risks?: number;
    counts?: { initial: TieredRiskCounts; final: TieredRiskCounts } | null;
  } | null;
  warnings: string[];
};

export type TieredCoverage = 'full' | 'partial' | 'unavailable';

// Outlook redesign (2026-07): the impersonal judgment on the stock…
export type TieredOutlook = 'bullish' | 'neutral' | 'bearish' | 'unknown';
// …and the personal instruction code derives from outlook × ownership.
export type TieredAction = 'enter' | 'keep_holding' | 'no_trade' | 'sell_all' | 'unknown';

// Warning-only earnings info: never gates anything, never moves numbers.
export type TieredEarnings = {
  next_date: string | null;
  days_until: number | null;
  warning_days: number;
  is_near: boolean;
  note: string | null;
};

// One display-only risk-card entry. The 13 ids and their value keys are
// fixed by the backend (risk_card.py); wording lives in the i18n layer.
export type TieredRiskCardEntry = {
  id: string;
  status: 'ok' | 'flag' | 'na';
  values: Record<string, unknown>;
};

// The deepest tier's verdict in summary form — what the run ends on.
export type TieredFinal = {
  tier: number;
  direction: 'buy' | 'hold' | 'sell' | 'unknown';
  outlook?: TieredOutlook;
  action?: TieredAction;
  coverage: 'full' | 'partial' | 'unavailable';
  confidence: string | null;
  levels: TieredLevels;
};

export type TieredTierSection = {
  tier: number;
  coverage: 'full' | 'partial' | 'unavailable';
  direction: 'buy' | 'hold' | 'sell' | 'unknown';
  confidence: string | null;
  score: number | null;
  levels: TieredLevels;
  narrative: string | null;
  warnings: string[];
  debate_detail?: TieredDebateDetail;
  risk_detail?: TieredRiskDetail;
};

// v2 slice 6: deterministic sizing block (share count or explicit refusal).
export type TieredSizing = {
  enabled: boolean;
  shares: number | null;
  // Multiplier keys died with tier 3 — absent on outlook-redesign runs.
  shares_before_multiplier?: number | null;
  risk_multiplier?: number | null;
  // Ownership block (absent on runs stored before the ownership input):
  // held shares, and the exit size a sell verdict prints from them.
  ownership?: number | null;
  sell_shares?: number | null;
  sell_shares_before_multiplier?: number | null;
  position_value: number | null;
  risk_amount: number | null;
  loss_per_share: number | null;
  lot_size: number;
  reason_code: string | null;
  refusal_reason: string | null;
  notes: string[];
  // Fee rate and the position cap were removed 2026-07-22 (they return
  // later as user inputs); old stored runs may still carry those keys.
  inputs: {
    capital: number | null;
    risk_fraction: number | null;
    // Absent on runs stored before the reward filter existed.
    reward_risk?: number | null;
    entry: number | null;
    stop_loss: number | null;
  };
};

export type TieredLlmUsage = {
  stages: Record<string, { calls: number; prompt_tokens: number; completion_tokens: number }>;
  total: { calls: number; prompt_tokens: number; completion_tokens: number };
  scope: string;
};

export type TieredResult = {
  symbol: string;
  market: string;
  tier: number;
  direction: 'buy' | 'hold' | 'sell' | 'unknown';
  score: number | null;
  confidence: string | null;
  coverage: 'full' | 'partial' | 'unavailable';
  levels: TieredLevels;
  levels_detail?: TieredLevelsDetail | null;
  narrative: string | null;
  warnings: string[];
  dimensions: TieredDimension[];
  signal: {
    logged: boolean;
    signal_id: number | null;
    created: boolean | null;
    reason: string | null;
  } | null;
  // v2 slice 6 additions — absent on old stored runs, so all optional.
  depth?: number;
  final?: TieredFinal | null;
  tier2?: TieredTierSection | null;
  tier3?: TieredTierSection | null;
  sizing?: TieredSizing | null;
  llm_usage?: TieredLlmUsage | null;
  // Outlook redesign additions — absent on old stored runs.
  outlook?: TieredOutlook;
  action?: TieredAction;
  earnings?: TieredEarnings | null;
  // Retired 2026-07-22 (kept so old stored runs still type-check).
  risk_card?: TieredRiskCardEntry[] | null;
  // Plan review (2026-07-22): per-column trade-plan warnings.
  plan_warnings?: TieredPlanWarnings | null;
};

export type TieredRunStatus = 'running' | 'done' | 'failed';

export type TieredRunSummary = {
  task_id: string;
  stock_code: string;
  status: TieredRunStatus;
  error: string | null;
  created_at: string | null;
  updated_at: string | null;
  // Digest of the stored report for the history rows. Null while running,
  // after a failure, or on rows stored before the backend sent these.
  // shares mirrors the report card: 0 = sizing ran but bought nothing,
  // null = the run has no sizing block (shown as a dash).
  direction?: 'buy' | 'hold' | 'sell' | 'unknown' | null;
  // Outlook digest: stored on new runs; the backend maps old runs'
  // buy/hold/sell to bullish/neutral/bearish so the filter is uniform.
  outlook?: TieredOutlook | null;
  shares?: number | null;
  // The tier the run went to (1-3); null while running/failed.
  tier?: number | null;
  // The sizing inputs the run used (capital in the ticker's own currency,
  // risk as a 0-1 fraction); null when the run had no sizing block.
  capital?: number | null;
  risk_fraction?: number | null;
  // The reward-to-risk goal the run planned toward (target = entry +
  // reward × risk); null on runs stored before it was recorded.
  reward_risk?: number | null;
};

export type TieredRun = TieredRunSummary & {
  result: TieredResult | null;
};

// Tier 3 retired (outlook redesign): the backend rejects depth 3 with a
// validation error and the alt form offers 1-2 only. The type keeps 3 so
// the legacy /tiered page (deliberately untouched) still compiles.
export type TieredDepth = 1 | 2 | 3;

export type TieredSizingRequest = {
  capital?: number;
  risk_fraction?: number;
  // Kept for API compatibility; the alt form no longer collects it.
  ownership?: number;
  // Reward-to-risk ratio the plan aims for (target = entry + R × risk).
  reward_risk?: number;
};

export const tieredApi = {
  start: async (
    stockCode: string,
    depth: TieredDepth = 1,
    sizing?: TieredSizingRequest,
  ): Promise<{ task_id: string }> => {
    const response = await apiClient.post<{ task_id: string }>('/api/v1/tiered/analyze', {
      stock_code: stockCode,
      depth,
      ...(sizing && Object.keys(sizing).length > 0 ? { sizing } : {}),
    });
    return response.data;
  },

  // 200 is the backend's cap; the page paginates client-side over this.
  listRuns: async (limit = 200): Promise<TieredRunSummary[]> => {
    const response = await apiClient.get<{ items: TieredRunSummary[] }>('/api/v1/tiered/runs', {
      params: { limit },
    });
    return response.data.items;
  },

  getRun: async (taskId: string): Promise<TieredRun> => {
    const response = await apiClient.get<TieredRun>(`/api/v1/tiered/runs/${taskId}`);
    return response.data;
  },

  // Saved sizing settings (.env-backed) — what a run uses when the form
  // sends nothing. Both null when no defaults are configured.
  sizingDefaults: async (): Promise<{
    capital: number | null;
    risk_fraction: number | null;
    reward_risk?: number | null;
  }> => {
    const response = await apiClient.get<{
      capital: number | null;
      risk_fraction: number | null;
      reward_risk?: number | null;
    }>('/api/v1/tiered/sizing-defaults');
    return response.data;
  },
};

File: api/v1/endpoints/tiered.py
# -*- coding: utf-8 -*-
"""Tiered-analysis API: run tier 1 for a symbol from the web UI.

A full run takes minutes (data fetch + LLM), so POST /analyze returns a
task id immediately; the run executes in a background thread. Runs
persist in the tiered_runs table (src/tiered_analysis/history.py), so
GET /runs serves a clickable history that survives page navigation and
server restarts, and GET /runs/{task_id} returns the stored full report.
"""
from __future__ import annotations

import logging
import threading
import uuid
from typing import Any, Dict, List, Optional

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, Field, field_validator

from src.tiered_analysis import history

logger = logging.getLogger(__name__)

router = APIRouter()


def _run_analysis(stock_code: str, depth: int = 1,
                  sizing_overrides: Optional[Dict[str, float]] = None,
                  hold_weeks: int = 2):
    """Indirection so tests can patch the multi-minute production run."""
    from src.tiered_analysis.integration import run_tiered_analysis

    return run_tiered_analysis(
        stock_code, depth=depth, sizing_overrides=sizing_overrides,
        hold_weeks=hold_weeks,
    )


class SizingOverride(BaseModel):
    """Per-run sizing inputs; saved settings fill whatever is omitted."""

    capital: Optional[float] = Field(default=None, gt=0)
    risk_fraction: Optional[float] = Field(default=None, gt=0, lt=1)
    #: Shares of this stock the user already holds (0 = none). Kept for
    #: API compatibility; the alt page no longer sends it (the ownership
    #: input is deferred to the future portfolio feature).
    ownership: Optional[int] = Field(default=None, ge=0)
    #: Target reward-to-risk ratio for the plan (target = entry + R × risk).
    reward_risk: Optional[float] = Field(default=None, gt=1, le=10)


class TieredAnalyzeRequest(BaseModel):
    stock_code: str
    #: 1 = the one-blob judge, 2 = the evidence vote. Tier 3 is retired
    #: (outlook redesign) — depth 3 is a validation error, not a clamp.
    depth: int = Field(default=1, ge=1, le=2)
    sizing: Optional[SizingOverride] = None
    #: Max hold time in weeks (2026-08-08): the horizon the AI judges
    #: against, shown on the report and used as the grading window.
    hold_weeks: int = Field(default=2, ge=1, le=4)
    #: Clock-gate override ("run anyway"): start during the trading day,
    #: analyzing the PREVIOUS completed session's close.
    run_anyway: bool = False

    @field_validator("stock_code")
    @classmethod
    def _not_blank(cls, value: str) -> str:
        cleaned = value.strip()
        if not cleaned:
            raise ValueError("stock_code must not be blank")
        return cleaned


def _serialize_levels(levels: Any) -> Dict[str, Any]:
    return {
        "entry": levels.entry,
        "secondary_entry": levels.secondary_entry,
        "stop_loss": levels.stop_loss,
        "take_profit": levels.take_profit,
    }


def _serialize_tier_section(report: Any) -> Optional[Dict[str, Any]]:
    """Tier 2/3 section: verdict + audit trail, no dimension duplication."""
    if report is None:
        return None
    section: Dict[str, Any] = {
        "tier": report.tier,
        "coverage": report.coverage.value,
        "direction": report.direction.value,
        "confidence": report.confidence,
        "score": report.score,
        "levels": _serialize_levels(report.levels),
        "narrative": report.narrative,
        "warnings": list(report.warnings),
    }
    if report.debate_detail is not None:
        section["debate_detail"] = report.debate_detail
    if report.risk_detail is not None:
        section["risk_detail"] = report.risk_detail
    return section


def _serialize_outcome(outcome: Any) -> Dict[str, Any]:
    report = outcome.report
    dimensions = []
    for dim in report.dimensions:
        dimensions.append({
            "dimension": dim.dimension,
            "kind": dim.kind.value,
            "coverage": dim.coverage.value,
            "is_actionable": dim.is_actionable,
            "payload": dim.payload,
            # UI formula receipts for derived metrics (technicals v2);
            # None on other dimensions and old runs.
            "formulas": dim.formulas,
            "narrative": dim.narrative,
            "warnings": list(dim.warnings),
            # The same notes keyed by the payload field ("group.key")
            # each is about; None on old stored runs.
            "field_notes": dim.field_notes,
            "citations": [
                {
                    "source_name": c.source_name,
                    "url": c.url,
                    "title": c.title,
                    "snippet": c.snippet,
                }
                for c in (dim.citations or [])
            ],
        })

    signal: Optional[Dict[str, Any]] = None
    if outcome.signal is not None:
        signal = {
            "logged": outcome.signal.logged,
            "signal_id": outcome.signal.signal_id,
            "created": outcome.signal.created,
            "reason": outcome.signal.reason,
        }

    state_reports = getattr(outcome.state, "reports", {}) or {}
    final = outcome.final_report or report
    return {
        "symbol": report.symbol,
        "market": report.market.value,
        "tier": report.tier,
        "direction": report.direction.value,
        "score": report.score,
        "confidence": report.confidence,
        "coverage": report.coverage.value,
        "levels": _serialize_levels(report.levels),
        "levels_detail": report.levels_detail,
        "narrative": report.narrative,
        "warnings": list(report.warnings),
        "dimensions": dimensions,
        "signal": signal,
        # v2 slice 6 (additive): depth, deeper-tier sections, sizing, cost.
        "depth": outcome.depth,
        "final": {
            "tier": final.tier,
            "direction": final.direction.value,
            "outlook": outcome.outlook.value,
            "action": outcome.action.value,
            "coverage": final.coverage.value,
            "confidence": final.confidence,
            "levels": _serialize_levels(final.levels),
        },
        "tier2": _serialize_tier_section(state_reports.get(2)),
        "sizing": outcome.sizing,
        "llm_usage": outcome.llm_usage,
        # Outlook redesign (additive): the impersonal judgment, the
        # personal action, and the earnings date (displayed on the
        # fundamentals card; kept here for old consumers).
        "outlook": outcome.outlook.value,
        "action": outcome.action.value,
        "earnings": outcome.earnings.to_detail() if outcome.earnings else None,
        # Retired 2026-07-22 — always None on new runs; old stored runs
        # still carry their card.
        "risk_card": outcome.risk_card,
        # Plan review (additive): structured per-column trade-plan
        # warnings — numbers only, the frontend words them.
        "plan_warnings": outcome.plan_warnings,
        # Max hold time (additive, 2026-08-08): the horizon in weeks the
        # AI was judged against; absent on old stored runs.
        "hold_weeks": getattr(outcome, "hold_weeks", None),
    }


def _run_task(task_id: str, stock_code: str, depth: int = 1,
              sizing_overrides: Optional[Dict[str, float]] = None,
              hold_weeks: int = 2) -> None:
    try:
        outcome = _run_analysis(stock_code, depth=depth,
                                sizing_overrides=sizing_overrides,
                                hold_weeks=hold_weeks)
        history.mark_done(task_id, _serialize_outcome(outcome))
    except Exception as exc:
        logger.error("tiered analysis task failed for %s: %s",
                     stock_code, exc, exc_info=True)
        history.mark_failed(task_id, str(exc))


def _effective_run_inputs(request: TieredAnalyzeRequest) -> Dict[str, Any]:
    """The settings the run will actually use — request overrides on top
    of saved .env values, with the web-form fallbacks underneath (the
    same resolution run_tiered_analysis performs) — recorded on the run
    row so history shows them while the run is still in flight."""
    from src.tiered_analysis.settings import (
        load_sizing_settings,
        merge_overrides,
        with_fallback_defaults,
    )

    sizing = request.sizing
    settings = with_fallback_defaults(merge_overrides(
        load_sizing_settings(),
        capital=sizing.capital if sizing else None,
        risk_fraction=sizing.risk_fraction if sizing else None,
        ownership=sizing.ownership if sizing else None,
        reward_risk=sizing.reward_risk if sizing else None,
    ))
    return {
        "tier": request.depth,
        "capital": settings.capital,
        "risk_fraction": settings.risk_fraction,
        "reward_risk": settings.reward_risk,
        "hold_weeks": request.hold_weeks,
    }


@router.post("/analyze", status_code=202)
def start_tiered_analysis(request: TieredAnalyzeRequest) -> Dict[str, Any]:
    """Kick off a tiered run in the background; returns a pollable task.

    CLOCK GATE (2026-08-08): runs are rejected (409) from market open
    until 30 minutes past the close in the exchange's own timezone — the
    app analyzes completed trading days only. ``run_anyway: true``
    overrides and analyzes the previous completed session instead.
    """
    from src.tiered_analysis.run_gate import clock_gate

    gate = clock_gate(request.stock_code, override=request.run_anyway)
    if gate.blocked:
        logger.info(
            "tiered run blocked by clock gate for %s (%s)",
            request.stock_code, gate.detail,
        )
        raise HTTPException(
            status_code=409,
            detail={"code": "market_open", "market": gate.market},
        )

    task_id = uuid.uuid4().hex
    history.create_run(task_id, request.stock_code,
                       inputs=_effective_run_inputs(request))
    sizing_overrides: Optional[Dict[str, float]] = None
    if request.sizing is not None:
        sizing_overrides = request.sizing.model_dump(exclude_none=True) or None
    worker = threading.Thread(
        target=_run_task,
        args=(task_id, request.stock_code, request.depth, sizing_overrides,
              request.hold_weeks),
        name=f"tiered-analysis-{request.stock_code}",
        daemon=True,
    )
    worker.start()
    return {"task_id": task_id, "stock_code": request.stock_code,
            "depth": request.depth, "status": "running"}


@router.get("/sizing-defaults")
def get_sizing_defaults() -> Dict[str, Optional[float]]:
    """Saved sizing settings (.env-backed) — capital and risk fraction —
    so the run form can show the values a run would use when the user
    provides none. Both are null when no defaults are configured."""
    from src.tiered_analysis.settings import load_sizing_settings

    settings = load_sizing_settings()
    return {"capital": settings.capital,
            "risk_fraction": settings.risk_fraction,
            "reward_risk": settings.reward_risk}


@router.get("/runs")
def list_tiered_runs(limit: int = 50) -> Dict[str, List[Dict[str, Any]]]:
    """Run history, newest first (summaries only)."""
    return {"items": history.list_runs(limit=limit)}


@router.get("/runs/{task_id}")
def get_tiered_run(task_id: str) -> Dict[str, Any]:
    """One run with its stored full report."""
    run = history.get_run(task_id)
    if run is None:
        raise HTTPException(status_code=404, detail="run not found")
    return run
